In [264]:
import pandas as pd

In [265]:
matches = pd.read_csv("matches.csv" , index_col = 0)

In [266]:
matches.head()

date   time            comp        round  day venue result   gf   ga  \
0  2024-08-17  12:30  Premier League  Matchweek 1  Sat  Away      W  2.0  0.0   
1  2024-08-25  16:30  Premier League  Matchweek 2  Sun  Home      W  2.0  0.0   
2  2024-09-01  16:00  Premier League  Matchweek 3  Sun  Away      W  3.0  0.0   
3  2024-09-14  15:00  Premier League  Matchweek 4  Sat  Home      L  0.0  1.0   
5  2024-09-21  15:00  Premier League  Matchweek 5  Sat  Home      W  3.0  0.0   

          opponent  ...  match report  notes    sh   sot  dist   fk pk pkatt  \
0     Ipswich Town  ...  Match Report    NaN  18.0   5.0  14.8  0.0  0     0   
1        Brentford  ...  Match Report    NaN  19.0   8.0  13.6  1.0  0     0   
2   Manchester Utd  ...  Match Report    NaN  11.0   3.0  13.4  0.0  0     0   
3  Nott'ham Forest  ...  Match Report    NaN  14.0   5.0  14.9  0.0  0     0   
5      Bournemouth  ...  Match Report    NaN  19.0  12.0  16.6  0.0  0     0   

  season       team  
0   2024  Liverpool  
1   2024  Liverpool  
2   2024  Liverpool  
3   2024  Liverpool  
5   2024  Liverpool  

[5 rows x 28 columns]

In [267]:
matches.shape

(3053, 28)

In [268]:
38 * 20 * 4

3040

In [269]:
matches["team"].value_counts()

team
Liverpool                   165
Manchester United           152
Everton                     152
Manchester City             152
Brighton and Hove Albion    152
Crystal Palace              152
West Ham United             152
Wolverhampton Wanderers     152
Newcastle United            152
Tottenham Hotspur           152
Aston Villa                 152
Arsenal                     152
Chelsea                     152
Fulham                      114
Brentford                   114
Burnley                     114
Leicester City              114
Leeds United                114
Southampton                 114
Bournemouth                  76
Nottingham Forest            76
Sheffield United             76
Norwich City                 38
Luton Town                   38
Watford                      38
West Bromwich Albion         38
Name: count, dtype: int64

In [270]:
matches[matches["team"] == "Liverpool"]

date   time            comp         round  day venue result   gf  \
0   2024-08-17  12:30  Premier League   Matchweek 1  Sat  Away      W  2.0   
1   2024-08-25  16:30  Premier League   Matchweek 2  Sun  Home      W  2.0   
2   2024-09-01  16:00  Premier League   Matchweek 3  Sun  Away      W  3.0   
3   2024-09-14  15:00  Premier League   Matchweek 4  Sat  Home      L  0.0   
5   2024-09-21  15:00  Premier League   Matchweek 5  Sat  Home      W  3.0   
..         ...    ...             ...           ...  ...   ...    ...  ...   
48  2021-05-08  20:15  Premier League  Matchweek 35  Sat  Home      W  2.0   
49  2021-05-13  20:15  Premier League  Matchweek 34  Thu  Away      W  4.0   
50  2021-05-16  16:30  Premier League  Matchweek 36  Sun  Away      W  2.0   
51  2021-05-19  20:15  Premier League  Matchweek 37  Wed  Away      W  3.0   
52  2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  2.0   

     ga         opponent  ...  match report  notes    sh   sot  dist   fk pk  \
0   0.0     Ipswich Town  ...  Match Report    NaN  18.0   5.0  14.8  0.0  0   
1   0.0        Brentford  ...  Match Report    NaN  19.0   8.0  13.6  1.0  0   
2   0.0   Manchester Utd  ...  Match Report    NaN  11.0   3.0  13.4  0.0  0   
3   1.0  Nott'ham Forest  ...  Match Report    NaN  14.0   5.0  14.9  0.0  0   
5   0.0      Bournemouth  ...  Match Report    NaN  19.0  12.0  16.6  0.0  0   
..  ...              ...  ...           ...    ...   ...   ...   ...  ... ..   
48  0.0      Southampton  ...  Match Report    NaN  14.0   6.0  13.6  0.0  0   
49  2.0   Manchester Utd  ...  Match Report    NaN  17.0   8.0  14.9  0.0  0   
50  1.0        West Brom  ...  Match Report    NaN  26.0   6.0  18.3  1.0  0   
51  0.0          Burnley  ...  Match Report    NaN  20.0   3.0  16.2  1.0  0   
52  0.0   Crystal Palace  ...  Match Report    NaN  19.0   5.0  14.7  1.0  0   

   pkatt season       team  
0      0   2024  Liverpool  
1      0   2024  Liverpool  
2      0   2024  Liverpool  
3      0   2024  Liverpool  
5      0   2024  Liverpool  
..   ...    ...        ...  
48     0   2021  Liverpool  
49     0   2021  Liverpool  
50     0   2021  Liverpool  
51     0   2021  Liverpool  
52     0   2021  Liverpool  

[165 rows x 28 columns]

In [271]:
matches.drop(columns=["notes"] , inplace = True)

In [272]:
matches["season"].value_counts()

season
2024    773
2023    760
2022    760
2021    760
Name: count, dtype: int64

In [273]:
matches["round"].value_counts()

round
Matchweek 1     81
Matchweek 8     81
Matchweek 2     81
Matchweek 12    81
Matchweek 11    81
Matchweek 10    81
Matchweek 9     81
Matchweek 13    81
Matchweek 7     81
Matchweek 5     81
Matchweek 4     81
Matchweek 6     81
Matchweek 3     81
Matchweek 33    80
Matchweek 28    80
Matchweek 30    80
Matchweek 31    80
Matchweek 32    80
Matchweek 35    80
Matchweek 29    80
Matchweek 36    80
Matchweek 37    80
Matchweek 34    80
Matchweek 26    80
Matchweek 27    80
Matchweek 21    80
Matchweek 18    80
Matchweek 25    80
Matchweek 24    80
Matchweek 23    80
Matchweek 22    80
Matchweek 20    80
Matchweek 19    80
Matchweek 17    80
Matchweek 16    80
Matchweek 15    80
Matchweek 14    80
Matchweek 38    80
Name: count, dtype: int64

In [274]:
matches.dtypes

date              object
time              object
comp              object
round             object
day               object
venue             object
result            object
gf               float64
ga               float64
opponent          object
xg               float64
xga              float64
poss             float64
attendance       float64
captain           object
formation         object
opp formation     object
referee           object
match report      object
sh               float64
sot              float64
dist             float64
fk               float64
pk                 int64
pkatt              int64
season             int64
team              object
dtype: object

In [275]:
##            Converting Date[String] into datetime

matches["date"] = pd.to_datetime(matches["date"])

In [276]:
matches.dtypes

date             datetime64[ns]
time                     object
comp                     object
round                    object
day                      object
venue                    object
result                   object
gf                      float64
ga                      float64
opponent                 object
xg                      float64
xga                     float64
poss                    float64
attendance              float64
captain                  object
formation                object
opp formation            object
referee                  object
match report             object
sh                      float64
sot                     float64
dist                    float64
fk                      float64
pk                        int64
pkatt                     int64
season                    int64
team                     object
dtype: object

In [277]:
## Creating Basic Predictors to train our machine learning model
## Creating a Predictor from veneus that will convert Home and Away matches into 1 and 0
matches["venue_code"] = matches["venue"].astype("category").cat.codes

In [278]:
matches

date   time            comp         round  day venue result   gf  \
0  2024-08-17  12:30  Premier League   Matchweek 1  Sat  Away      W  2.0   
1  2024-08-25  16:30  Premier League   Matchweek 2  Sun  Home      W  2.0   
2  2024-09-01  16:00  Premier League   Matchweek 3  Sun  Away      W  3.0   
3  2024-09-14  15:00  Premier League   Matchweek 4  Sat  Home      L  0.0   
5  2024-09-21  15:00  Premier League   Matchweek 5  Sat  Home      W  3.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga         opponent  ...  match report    sh   sot  dist   fk pk pkatt  \
0   0.0     Ipswich Town  ...  Match Report  18.0   5.0  14.8  0.0  0     0   
1   0.0        Brentford  ...  Match Report  19.0   8.0  13.6  1.0  0     0   
2   0.0   Manchester Utd  ...  Match Report  11.0   3.0  13.4  0.0  0     0   
3   1.0  Nott'ham Forest  ...  Match Report  14.0   5.0  14.9  0.0  0     0   
5   0.0      Bournemouth  ...  Match Report  19.0  12.0  16.6  0.0  0     0   
..  ...              ...  ...           ...   ...   ...   ...  ... ..   ...   
38  4.0        Tottenham  ...  Match Report   8.0   1.0  18.2  0.0  0     0   
39  2.0   Crystal Palace  ...  Match Report   7.0   0.0  13.4  1.0  0     0   
40  0.0          Everton  ...  Match Report  10.0   3.0  18.5  0.0  0     0   
41  1.0    Newcastle Utd  ...  Match Report  11.0   1.0  18.3  1.0  0     0   
42  0.0          Burnley  ...  Match Report  12.0   3.0  17.4  0.0  0     0   

   season              team  venue_code  
0    2024         Liverpool           0  
1    2024         Liverpool           1  
2    2024         Liverpool           0  
3    2024         Liverpool           1  
5    2024         Liverpool           1  
..    ...               ...         ...  
38   2021  Sheffield United           0  
39   2021  Sheffield United           1  
40   2021  Sheffield United           0  
41   2021  Sheffield United           0  
42   2021  Sheffield United           1  

[3053 rows x 28 columns]

In [279]:
## Creating a unique code for each opponent squad
matches["opp_code"] = matches["opponent"].astype("category").cat.codes

In [280]:
matches

date   time            comp         round  day venue result   gf  \
0  2024-08-17  12:30  Premier League   Matchweek 1  Sat  Away      W  2.0   
1  2024-08-25  16:30  Premier League   Matchweek 2  Sun  Home      W  2.0   
2  2024-09-01  16:00  Premier League   Matchweek 3  Sun  Away      W  3.0   
3  2024-09-14  15:00  Premier League   Matchweek 4  Sat  Home      L  0.0   
5  2024-09-21  15:00  Premier League   Matchweek 5  Sat  Home      W  3.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga         opponent  ...    sh   sot  dist   fk pk pkatt season  \
0   0.0     Ipswich Town  ...  18.0   5.0  14.8  0.0  0     0   2024   
1   0.0        Brentford  ...  19.0   8.0  13.6  1.0  0     0   2024   
2   0.0   Manchester Utd  ...  11.0   3.0  13.4  0.0  0     0   2024   
3   1.0  Nott'ham Forest  ...  14.0   5.0  14.9  0.0  0     0   2024   
5   0.0      Bournemouth  ...  19.0  12.0  16.6  0.0  0     0   2024   
..  ...              ...  ...   ...   ...   ...  ... ..   ...    ...   
38  4.0        Tottenham  ...   8.0   1.0  18.2  0.0  0     0   2021   
39  2.0   Crystal Palace  ...   7.0   0.0  13.4  1.0  0     0   2021   
40  0.0          Everton  ...  10.0   3.0  18.5  0.0  0     0   2021   
41  1.0    Newcastle Utd  ...  11.0   1.0  18.3  1.0  0     0   2021   
42  0.0          Burnley  ...  12.0   3.0  17.4  0.0  0     0   2021   

                team venue_code  opp_code  
0          Liverpool          0        10  
1          Liverpool          1         3  
2          Liverpool          0        16  
3          Liverpool          1        19  
5          Liverpool          1         2  
..               ...        ...       ...  
38  Sheffield United          0        22  
39  Sheffield United          1         7  
40  Sheffield United          0         8  
41  Sheffield United          0        17  
42  Sheffield United          1         5  

[3053 rows x 29 columns]

In [281]:
## Creating a new column for time
matches["hour"] = matches["time"].str.replace(":.+" , "", regex = True).astype("int")

In [282]:
## Creating a column for day of the week where the matches are the played
matches["day_code"] = matches["date"].dt.dayofweek

In [283]:
matches

date   time            comp         round  day venue result   gf  \
0  2024-08-17  12:30  Premier League   Matchweek 1  Sat  Away      W  2.0   
1  2024-08-25  16:30  Premier League   Matchweek 2  Sun  Home      W  2.0   
2  2024-09-01  16:00  Premier League   Matchweek 3  Sun  Away      W  3.0   
3  2024-09-14  15:00  Premier League   Matchweek 4  Sat  Home      L  0.0   
5  2024-09-21  15:00  Premier League   Matchweek 5  Sat  Home      W  3.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga         opponent  ...  dist   fk  pk  pkatt season              team  \
0   0.0     Ipswich Town  ...  14.8  0.0   0      0   2024         Liverpool   
1   0.0        Brentford  ...  13.6  1.0   0      0   2024         Liverpool   
2   0.0   Manchester Utd  ...  13.4  0.0   0      0   2024         Liverpool   
3   1.0  Nott'ham Forest  ...  14.9  0.0   0      0   2024         Liverpool   
5   0.0      Bournemouth  ...  16.6  0.0   0      0   2024         Liverpool   
..  ...              ...  ...   ...  ...  ..    ...    ...               ...   
38  4.0        Tottenham  ...  18.2  0.0   0      0   2021  Sheffield United   
39  2.0   Crystal Palace  ...  13.4  1.0   0      0   2021  Sheffield United   
40  0.0          Everton  ...  18.5  0.0   0      0   2021  Sheffield United   
41  1.0    Newcastle Utd  ...  18.3  1.0   0      0   2021  Sheffield United   
42  0.0          Burnley  ...  17.4  0.0   0      0   2021  Sheffield United   

   venue_code opp_code hour  day_code  
0           0       10   12         5  
1           1        3   16         6  
2           0       16   16         6  
3           1       19   15         5  
5           1        2   15         5  
..        ...      ...  ...       ...  
38          0       22   19         6  
39          1        7   15         5  
40          0        8   19         6  
41          0       17   18         2  
42          1        5   16         6  

[3053 rows x 31 columns]

In [284]:
## Creaating a column for Target for the results [0 when team LOST & 1 when team WON]  
matches["target"] = (matches["result"] == "W").astype("int")

In [285]:
matches

date   time            comp         round  day venue result   gf  \
0  2024-08-17  12:30  Premier League   Matchweek 1  Sat  Away      W  2.0   
1  2024-08-25  16:30  Premier League   Matchweek 2  Sun  Home      W  2.0   
2  2024-09-01  16:00  Premier League   Matchweek 3  Sun  Away      W  3.0   
3  2024-09-14  15:00  Premier League   Matchweek 4  Sat  Home      L  0.0   
5  2024-09-21  15:00  Premier League   Matchweek 5  Sat  Home      W  3.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga         opponent  ...   fk  pk  pkatt  season              team  \
0   0.0     Ipswich Town  ...  0.0   0      0    2024         Liverpool   
1   0.0        Brentford  ...  1.0   0      0    2024         Liverpool   
2   0.0   Manchester Utd  ...  0.0   0      0    2024         Liverpool   
3   1.0  Nott'ham Forest  ...  0.0   0      0    2024         Liverpool   
5   0.0      Bournemouth  ...  0.0   0      0    2024         Liverpool   
..  ...              ...  ...  ...  ..    ...     ...               ...   
38  4.0        Tottenham  ...  0.0   0      0    2021  Sheffield United   
39  2.0   Crystal Palace  ...  1.0   0      0    2021  Sheffield United   
40  0.0          Everton  ...  0.0   0      0    2021  Sheffield United   
41  1.0    Newcastle Utd  ...  1.0   0      0    2021  Sheffield United   
42  0.0          Burnley  ...  0.0   0      0    2021  Sheffield United   

   venue_code opp_code hour day_code  target  
0           0       10   12        5       1  
1           1        3   16        6       1  
2           0       16   16        6       1  
3           1       19   15        5       0  
5           1        2   15        5       1  
..        ...      ...  ...      ...     ...  
38          0       22   19        6       0  
39          1        7   15        5       0  
40          0        8   19        6       1  
41          0       17   18        2       0  
42          1        5   16        6       1  

[3053 rows x 32 columns]

In [286]:
## Creating machine learning model
from sklearn.ensemble import RandomForestClassifier

In [287]:
rf = RandomForestClassifier(n_estimators = 50, min_samples_split = 10 , random_state = 1)

In [288]:
## Trainin the data on the matches that are played before the year I want to predict
train = matches[matches["date"] < '2024-01-01']

In [289]:
## Testing the data on which we are going to make predictions
test = matches[matches["date"] > '2024-01-01']

In [290]:
predictors = ["venue_code" , "opp_code" , "hour" , "day_code"]

In [291]:
rf.fit(train[predictors] , train["target"])

RandomForestClassifier(min_samples_split=10, n_estimators=50, random_state=1)

In [292]:
preds = rf.predict(test[predictors])

In [293]:
## If I predicted the win what is percentage on which occasions prediction accurate
from sklearn.metrics import accuracy_score

In [294]:
acc = accuracy_score(test["target"], preds)

In [295]:
acc

0.6517150395778364

In [296]:
## Checing the situations where the ccuracy of the model was HIGH vs LOW
combined = pd.DataFrame(dict(actual=test["target"] , prediction = preds))

In [297]:
## Ceating a table which shows what acctually was the prediction
pd.crosstab(index = combined["actual"] , columns = combined["prediction"])

prediction    0   1
actual             
0           185  46
1            86  62

--We predicted Losses/Draws 185 times and we were correct but only 86 times we were wrong.
--We predicted Wins 46 times and the actual wins were 62 means our prediction was way off than the actual data.

In [299]:
## Using another metric to see what is the prediction score
from sklearn.metrics import precision_score

In [300]:
precision_score(test["target"], preds)

0.5740740740740741

In [301]:
## Creating a DataFrame for every Squad in our Data
grouped_matches = matches.groupby("team")

In [302]:
group = grouped_matches.get_group("Liverpool")

In [303]:
group

date   time            comp         round  day venue result   gf  \
0  2024-08-17  12:30  Premier League   Matchweek 1  Sat  Away      W  2.0   
1  2024-08-25  16:30  Premier League   Matchweek 2  Sun  Home      W  2.0   
2  2024-09-01  16:00  Premier League   Matchweek 3  Sun  Away      W  3.0   
3  2024-09-14  15:00  Premier League   Matchweek 4  Sat  Home      L  0.0   
5  2024-09-21  15:00  Premier League   Matchweek 5  Sat  Home      W  3.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
48 2021-05-08  20:15  Premier League  Matchweek 35  Sat  Home      W  2.0   
49 2021-05-13  20:15  Premier League  Matchweek 34  Thu  Away      W  4.0   
50 2021-05-16  16:30  Premier League  Matchweek 36  Sun  Away      W  2.0   
51 2021-05-19  20:15  Premier League  Matchweek 37  Wed  Away      W  3.0   
52 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  2.0   

     ga         opponent  ...   fk  pk  pkatt  season       team venue_code  \
0   0.0     Ipswich Town  ...  0.0   0      0    2024  Liverpool          0   
1   0.0        Brentford  ...  1.0   0      0    2024  Liverpool          1   
2   0.0   Manchester Utd  ...  0.0   0      0    2024  Liverpool          0   
3   1.0  Nott'ham Forest  ...  0.0   0      0    2024  Liverpool          1   
5   0.0      Bournemouth  ...  0.0   0      0    2024  Liverpool          1   
..  ...              ...  ...  ...  ..    ...     ...        ...        ...   
48  0.0      Southampton  ...  0.0   0      0    2021  Liverpool          1   
49  2.0   Manchester Utd  ...  0.0   0      0    2021  Liverpool          0   
50  1.0        West Brom  ...  1.0   0      0    2021  Liverpool          0   
51  0.0          Burnley  ...  1.0   0      0    2021  Liverpool          0   
52  0.0   Crystal Palace  ...  1.0   0      0    2021  Liverpool          1   

   opp_code hour day_code  target  
0        10   12        5       1  
1         3   16        6       1  
2        16   16        6       1  
3        19   15        5       0  
5         2   15        5       1  
..      ...  ...      ...     ...  
48       21   20        5       1  
49       16   20        3       1  
50       24   16        6       1  
51        5   20        2       1  
52        7   16        6       1  

[165 rows x 32 columns]

In [304]:
## Computing Rolling Averages [We can use how Liverpool performed in their previous 3 or 4 mathches and then the model can train on that specific data]
def rolling_averages(group, cols, new_cols):
    group = group.sort_values("date")
    rolling_stats = group[cols].rolling(5, closed='left').mean()
    group[new_cols] = rolling_stats
    group = group.dropna(subset = new_cols)
    return group

In [305]:
## Adding some new columns to calculate rolling averages
cols = ["gf" , "ga" , "sh" , "sot" , "dist" , "fk" , "pk" , "pkatt"]
new_cols = [f"{c}_rolling" for c in cols]

In [306]:
new_cols

['gf_rolling',
 'ga_rolling',
 'sh_rolling',
 'sot_rolling',
 'dist_rolling',
 'fk_rolling',
 'pk_rolling',
 'pkatt_rolling']

In [307]:
rolling_averages(group, cols, new_cols)

date   time            comp         round  day venue result   gf  \
9  2020-10-24  20:00  Premier League   Matchweek 6  Sat  Home      W  2.0   
11 2020-10-31  17:30  Premier League   Matchweek 7  Sat  Home      W  2.0   
13 2020-11-08  16:30  Premier League   Matchweek 8  Sun  Away      D  1.0   
14 2020-11-22  19:15  Premier League   Matchweek 9  Sun  Home      W  3.0   
16 2020-11-28  12:30  Premier League  Matchweek 10  Sat  Away      D  1.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
12 2024-10-27  16:30  Premier League   Matchweek 9  Sun  Away      D  2.0   
14 2024-11-02  15:00  Premier League  Matchweek 10  Sat  Home      W  2.0   
16 2024-11-09  20:00  Premier League  Matchweek 11  Sat  Home      W  2.0   
17 2024-11-24  14:00  Premier League  Matchweek 12  Sun  Away      W  3.0   
19 2024-12-01  16:00  Premier League  Matchweek 13  Sun  Home      W  2.0   

     ga         opponent  ...  day_code  target  gf_rolling  ga_rolling  \
9   1.0    Sheffield Utd  ...         5       1         2.6         2.6   
11  1.0         West Ham  ...         5       1         2.2         2.2   
13  1.0  Manchester City  ...         6       0         2.2         2.4   
14  0.0   Leicester City  ...         6       1         1.8         2.4   
16  1.0         Brighton  ...         5       0         2.0         1.0   
..  ...              ...  ...       ...     ...         ...         ...   
12  2.0          Arsenal  ...         6       0         1.6         0.6   
14  1.0         Brighton  ...         5       1         2.0         0.8   
16  0.0      Aston Villa  ...         5       1         1.8         1.0   
17  2.0      Southampton  ...         6       1         1.8         0.8   
19  0.0  Manchester City  ...         6       1         2.2         1.2   

   sh_rolling sot_rolling dist_rolling fk_rolling pk_rolling  pkatt_rolling  
9        19.0         6.6        17.36        0.8        0.4            0.4  
11       18.4         6.8        17.44        1.0        0.0            0.0  
13       16.4         6.6        17.80        1.0        0.2            0.2  
14       14.0         5.4        18.64        1.2        0.4            0.4  
16       16.0         6.0        17.92        1.0        0.4            0.4  
..        ...         ...          ...        ...        ...            ...  
12       13.2         6.0        18.22        0.4        0.4            0.4  
14       12.2         5.8        18.80        0.4        0.4            0.4  
16       11.6         5.0        18.28        0.4        0.4            0.4  
17       12.6         5.0        17.90        0.2        0.2            0.2  
19       14.6         6.2        17.76        0.4        0.4            0.4  

[160 rows x 40 columns]

In [308]:
matches_rolling = matches.groupby("team").apply(lambda x: rolling_averages(x, cols, new_cols))

C:\Users\muham\AppData\Local\Temp\ipykernel_2336\4052147919.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  matches_rolling = matches.groupby("team").apply(lambda x: rolling_averages(x, cols, new_cols))


In [309]:
matches_rolling

date   time            comp         round  \
team                                                                         
Arsenal                 9  2020-10-25  19:15  Premier League   Matchweek 6   
                        11 2020-11-01  16:30  Premier League   Matchweek 7   
                        13 2020-11-08  19:15  Premier League   Matchweek 8   
                        14 2020-11-22  16:30  Premier League   Matchweek 9   
                        16 2020-11-29  19:15  Premier League  Matchweek 10   
...                               ...    ...             ...           ...   
Wolverhampton Wanderers 40 2024-04-24  19:45  Premier League  Matchweek 29   
                        41 2024-04-27  15:00  Premier League  Matchweek 35   
                        42 2024-05-04  17:30  Premier League  Matchweek 36   
                        43 2024-05-11  15:00  Premier League  Matchweek 37   
                        44 2024-05-19  16:00  Premier League  Matchweek 38   

                            day venue result   gf   ga         opponent  ...  \
team                                                                     ...   
Arsenal                 9   Sun  Home      L  0.0  1.0   Leicester City  ...   
                        11  Sun  Away      W  1.0  0.0   Manchester Utd  ...   
                        13  Sun  Home      L  0.0  3.0      Aston Villa  ...   
                        14  Sun  Away      D  0.0  0.0     Leeds United  ...   
                        16  Sun  Home      L  1.0  2.0           Wolves  ...   
...                         ...   ...    ...  ...  ...              ...  ...   
Wolverhampton Wanderers 40  Wed  Home      L  0.0  1.0      Bournemouth  ...   
                        41  Sat  Home      W  2.0  1.0       Luton Town  ...   
                        42  Sat  Away      L  1.0  5.0  Manchester City  ...   
                        43  Sat  Home      L  1.0  3.0   Crystal Palace  ...   
                        44  Sun  Away      L  0.0  2.0        Liverpool  ...   

                            day_code  target  gf_rolling  ga_rolling  \
team                                                                   
Arsenal                 9          6       0         1.6         1.2   
                        11         6       1         1.0         1.4   
                        13         6       0         0.8         1.2   
                        14         6       0         0.6         1.2   
                        16         6       0         0.2         1.0   
...                              ...     ...         ...         ...   
Wolverhampton Wanderers 40         2       0         0.8         1.8   
                        41         5       1         0.8         1.6   
                        42         5       0         1.0         1.6   
                        43         5       0         1.0         2.2   
                        44         6       0         0.8         2.4   

                           sh_rolling sot_rolling dist_rolling fk_rolling  \
team                                                                        
Arsenal                 9         8.2         3.8        15.72        0.8   
                        11        8.0         3.6        15.90        0.6   
                        13        7.8         3.2        17.86        0.6   
                        14        9.6         3.0        18.14        0.6   
                        16       10.2         2.4        18.80        1.0   
...                               ...         ...          ...        ...   
Wolverhampton Wanderers 40       10.0         3.4        17.72        0.4   
                        41       10.4         3.6        18.82        0.2   
                        42       11.4         4.2        20.10        0.2   
                        43        9.2         3.2        16.82        0.0   
                        44        9.8         4.2        16.52        0.0   

                           pk_rolling 

In [310]:
matches_rolling = matches_rolling.droplevel('team')

In [311]:
matches_rolling

date   time            comp         round  day venue result   gf  \
9  2020-10-25  19:15  Premier League   Matchweek 6  Sun  Home      L  0.0   
11 2020-11-01  16:30  Premier League   Matchweek 7  Sun  Away      W  1.0   
13 2020-11-08  19:15  Premier League   Matchweek 8  Sun  Home      L  0.0   
14 2020-11-22  16:30  Premier League   Matchweek 9  Sun  Away      D  0.0   
16 2020-11-29  19:15  Premier League  Matchweek 10  Sun  Home      L  1.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
40 2024-04-24  19:45  Premier League  Matchweek 29  Wed  Home      L  0.0   
41 2024-04-27  15:00  Premier League  Matchweek 35  Sat  Home      W  2.0   
42 2024-05-04  17:30  Premier League  Matchweek 36  Sat  Away      L  1.0   
43 2024-05-11  15:00  Premier League  Matchweek 37  Sat  Home      L  1.0   
44 2024-05-19  16:00  Premier League  Matchweek 38  Sun  Away      L  0.0   

     ga         opponent  ...  day_code  target  gf_rolling  ga_rolling  \
9   1.0   Leicester City  ...         6       0         1.6         1.2   
11  0.0   Manchester Utd  ...         6       1         1.0         1.4   
13  3.0      Aston Villa  ...         6       0         0.8         1.2   
14  0.0     Leeds United  ...         6       0         0.6         1.2   
16  2.0           Wolves  ...         6       0         0.2         1.0   
..  ...              ...  ...       ...     ...         ...         ...   
40  1.0      Bournemouth  ...         2       0         0.8         1.8   
41  1.0       Luton Town  ...         5       1         0.8         1.6   
42  5.0  Manchester City  ...         5       0         1.0         1.6   
43  3.0   Crystal Palace  ...         5       0         1.0         2.2   
44  2.0        Liverpool  ...         6       0         0.8         2.4   

   sh_rolling sot_rolling dist_rolling fk_rolling pk_rolling  pkatt_rolling  
9         8.2         3.8        15.72        0.8        0.0            0.0  
11        8.0         3.6        15.90        0.6        0.0            0.0  
13        7.8         3.2        17.86        0.6        0.2            0.2  
14        9.6         3.0        18.14        0.6        0.2            0.2  
16       10.2         2.4        18.80        1.0        0.2            0.2  
..        ...         ...          ...        ...        ...            ...  
40       10.0         3.4        17.72        0.4        0.2            0.2  
41       10.4         3.6        18.82        0.2        0.2            0.2  
42       11.4         4.2        20.10        0.2        0.2            0.2  
43        9.2         3.2        16.82        0.0        0.0            0.0  
44        9.8         4.2        16.52        0.0        0.0            0.0  

[2918 rows x 40 columns]

In [312]:
matches_rolling.index = range(matches_rolling.shape[0])

In [313]:
matches_rolling

date   time            comp         round  day venue result   gf  \
0    2020-10-25  19:15  Premier League   Matchweek 6  Sun  Home      L  0.0   
1    2020-11-01  16:30  Premier League   Matchweek 7  Sun  Away      W  1.0   
2    2020-11-08  19:15  Premier League   Matchweek 8  Sun  Home      L  0.0   
3    2020-11-22  16:30  Premier League   Matchweek 9  Sun  Away      D  0.0   
4    2020-11-29  19:15  Premier League  Matchweek 10  Sun  Home      L  1.0   
...         ...    ...             ...           ...  ...   ...    ...  ...   
2913 2024-04-24  19:45  Premier League  Matchweek 29  Wed  Home      L  0.0   
2914 2024-04-27  15:00  Premier League  Matchweek 35  Sat  Home      W  2.0   
2915 2024-05-04  17:30  Premier League  Matchweek 36  Sat  Away      L  1.0   
2916 2024-05-11  15:00  Premier League  Matchweek 37  Sat  Home      L  1.0   
2917 2024-05-19  16:00  Premier League  Matchweek 38  Sun  Away      L  0.0   

       ga         opponent  ...  day_code  target  gf_rolling  ga_rolling  \
0     1.0   Leicester City  ...         6       0         1.6         1.2   
1     0.0   Manchester Utd  ...         6       1         1.0         1.4   
2     3.0      Aston Villa  ...         6       0         0.8         1.2   
3     0.0     Leeds United  ...         6       0         0.6         1.2   
4     2.0           Wolves  ...         6       0         0.2         1.0   
...   ...              ...  ...       ...     ...         ...         ...   
2913  1.0      Bournemouth  ...         2       0         0.8         1.8   
2914  1.0       Luton Town  ...         5       1         0.8         1.6   
2915  5.0  Manchester City  ...         5       0         1.0         1.6   
2916  3.0   Crystal Palace  ...         5       0         1.0         2.2   
2917  2.0        Liverpool  ...         6       0         0.8         2.4   

     sh_rolling sot_rolling dist_rolling fk_rolling pk_rolling  pkatt_rolling  
0           8.2         3.8        15.72        0.8        0.0            0.0  
1           8.0         3.6        15.90        0.6        0.0            0.0  
2           7.8         3.2        17.86        0.6        0.2            0.2  
3           9.6         3.0        18.14        0.6        0.2            0.2  
4          10.2         2.4        18.80        1.0        0.2            0.2  
...         ...         ...          ...        ...        ...            ...  
2913       10.0         3.4        17.72        0.4        0.2            0.2  
2914       10.4         3.6        18.82        0.2        0.2            0.2  
2915       11.4         4.2        20.10        0.2        0.2            0.2  
2916        9.2         3.2        16.82        0.0        0.0            0.0  
2917        9.8         4.2        16.52        0.0        0.0            0.0  

[2918 rows x 40 columns]

In [314]:
def make_predictions(data, predictors):
    # Split data into train and test sets
    test = data[data["date"] > '2024-01-01']
    train = data[data["date"] < '2024-01-01']

    # Ensure predictors exist in the train dataset
    missing_cols = [col for col in predictors if col not in train.columns]
    if missing_cols:
        raise ValueError(f"The following predictors are missing from the dataset: {missing_cols}")
    
    # Fit the model
    rf.fit(train[predictors], train["target"])

    # Predict on the test set
    preds = rf.predict(test[predictors])

    # Combine predictions and actual results
    combined = pd.DataFrame(dict(actual=test["target"], predicted=preds), index=test.index)

    # Calculate precision
    precision = precision_score(test["target"], preds)

    return combined, precision


In [315]:
combined, precision = make_predictions(matches_rolling, predictors + new_cols)

In [316]:
precision

0.6120689655172413

In [317]:
combined

actual  predicted
129        1          1
130        1          0
131        1          0
132        1          1
133        1          1
...      ...        ...
2913       0          0
2914       1          0
2915       0          0
2916       0          0
2917       0          0

[379 rows x 2 columns]

In [318]:
combined = combined.merge(matches_rolling[["date", "team", "opponent", "result"]], left_index = True, right_index = True)

In [319]:
combined

actual  predicted       date                     team         opponent  \
129        1          1 2024-01-20                  Arsenal   Crystal Palace   
130        1          0 2024-01-30                  Arsenal  Nott'ham Forest   
131        1          0 2024-02-04                  Arsenal        Liverpool   
132        1          1 2024-02-11                  Arsenal         West Ham   
133        1          1 2024-02-17                  Arsenal          Burnley   
...      ...        ...        ...                      ...              ...   
2913       0          0 2024-04-24  Wolverhampton Wanderers      Bournemouth   
2914       1          0 2024-04-27  Wolverhampton Wanderers       Luton Town   
2915       0          0 2024-05-04  Wolverhampton Wanderers  Manchester City   
2916       0          0 2024-05-11  Wolverhampton Wanderers   Crystal Palace   
2917       0          0 2024-05-19  Wolverhampton Wanderers        Liverpool   

     result  
129       W  
130       W  
131       W  
132       W  
133       W  
...     ...  
2913      L  
2914      W  
2915      L  
2916      L  
2917      L  

[379 rows x 6 columns]

In [320]:
## Mapping the missmatches names in the data  with the correct names in both opponent and team columns for  better predicions
class MissingDict(dict):
    __missing__ = lambda self, key: key

maps_values = {
    "Brighton & Hove Albion": "Brighton",
    "Manchester City": "Man City",
    "Manchester United": "Man Utd",
    "Newcastle United": "Newcastle",
    "Nottingham Forest": "Nottm Forest",
    "Sheffield United": "Sheffield Utd",
    "Tottenham Hotspur": "Tottenham",
    "West Ham United": "West Ham",
    "Wolverhampton Wanderers": "Wolves"
}
mapping = MissingDict(**maps_values)

In [321]:
mapping["West Ham United"]

'West Ham'

In [322]:
combined["new_team"] = combined["team"].map(mapping)

In [323]:
combined

actual  predicted       date                     team         opponent  \
129        1          1 2024-01-20                  Arsenal   Crystal Palace   
130        1          0 2024-01-30                  Arsenal  Nott'ham Forest   
131        1          0 2024-02-04                  Arsenal        Liverpool   
132        1          1 2024-02-11                  Arsenal         West Ham   
133        1          1 2024-02-17                  Arsenal          Burnley   
...      ...        ...        ...                      ...              ...   
2913       0          0 2024-04-24  Wolverhampton Wanderers      Bournemouth   
2914       1          0 2024-04-27  Wolverhampton Wanderers       Luton Town   
2915       0          0 2024-05-04  Wolverhampton Wanderers  Manchester City   
2916       0          0 2024-05-11  Wolverhampton Wanderers   Crystal Palace   
2917       0          0 2024-05-19  Wolverhampton Wanderers        Liverpool   

     result new_team  
129       W  Arsenal  
130       W  Arsenal  
131       W  Arsenal  
132       W  Arsenal  
133       W  Arsenal  
...     ...      ...  
2913      L   Wolves  
2914      W   Wolves  
2915      L   Wolves  
2916      L   Wolves  
2917      L   Wolves  

[379 rows x 7 columns]

In [324]:
merged = combined.merge(combined, left_on = ["date" , "new_team"], right_on = ["date" , "opponent"])

In [325]:
merged

actual_x  predicted_x       date                   team_x  \
0           1            1 2024-01-20                  Arsenal   
1           1            0 2024-01-30                  Arsenal   
2           1            0 2024-02-04                  Arsenal   
3           1            1 2024-02-11                  Arsenal   
4           1            1 2024-02-17                  Arsenal   
..        ...          ...        ...                      ...   
269         0            0 2024-04-24  Wolverhampton Wanderers   
270         1            0 2024-04-27  Wolverhampton Wanderers   
271         0            0 2024-05-04  Wolverhampton Wanderers   
272         0            0 2024-05-11  Wolverhampton Wanderers   
273         0            0 2024-05-19  Wolverhampton Wanderers   

          opponent_x result_x new_team_x  actual_y  predicted_y  \
0     Crystal Palace        W    Arsenal         0            0   
1    Nott'ham Forest        W    Arsenal         0            0   
2          Liverpool        W    Arsenal         0            1   
3           West Ham        W    Arsenal         0            0   
4            Burnley        W    Arsenal         0            0   
..               ...      ...        ...       ...          ...   
269      Bournemouth        L     Wolves         1            0   
270       Luton Town        W     Wolves         0            0   
271  Manchester City        L     Wolves         1            1   
272   Crystal Palace        L     Wolves         1            1   
273        Liverpool        L     Wolves         1            1   

                team_y opponent_y result_y      new_team_y  
0       Crystal Palace    Arsenal        L  Crystal Palace  
1    Nottingham Forest    Arsenal        L    Nottm Forest  
2            Liverpool    Arsenal        L       Liverpool  
3      West Ham United    Arsenal        L        West Ham  
4              Burnley    Arsenal        L         Burnley  
..                 ...        ...      ...             ...  
269        Bournemouth     Wolves        W     Bournemouth  
270         Luton Town     Wolves        L      Luton Town  
271    Manchester City     Wolves        W        Man City  
272     Crystal Palace     Wolves        W  Crystal Palace  
273          Liverpool     Wolves        W       Liverpool  

[274 rows x 13 columns]

In [326]:
merged[(merged["predicted_x"]==1) & (merged["predicted_y"]==0)]["actual_x"].value_counts()

actual_x
1    38
0    28
Name: count, dtype: int64

In [327]:
## Dividing our Prediction with the actual number of matches played in the preimer league
38/40

0.95

In [328]:
import pandas as pd

# Assuming 'actual_x' is your Series containing the counts of correct and incorrect predictions
actual_x = pd.Series([38, 28], index=[1, 0], name='count')

# Total matches (sum of all counts)
total_matches = actual_x.sum()

# Number of correct predictions for positive cases
correct_positive_predictions = actual_x[1]  # The count of '1' predictions

# Calculate the accuracy for positive cases
accuracy = correct_positive_predictions / total_matches

# Display the accuracy as a percentage
print(f"Accuracy: {accuracy:.2%}")


Accuracy: 57.58%


In [329]:
!pip install xgboost

In [330]:
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [375]:
matches = pd.read_csv("matches.csv")

In [360]:
from sklearn.preprocessing import LabelEncoder

matches['venue_code'] = LabelEncoder().fit_transform(matches['venue'])
matches['opp_code'] = LabelEncoder().fit_transform(matches['opponent'])
matches['hour'] = pd.to_datetime(matches['time']).dt.hour
# Mapping for abbreviated day names
day_mapping = {
    'Mon': 0, 'Tue': 1, 'Wed': 2, 'Thu': 3,
    'Fri': 4, 'Sat': 5, 'Sun': 6
}

# Apply mapping to create day_code
matches['day_code'] = matches['day'].map(day_mapping)

C:\Users\muham\AppData\Local\Temp\ipykernel_2336\3431187377.py:5: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  matches['hour'] = pd.to_datetime(matches['time']).dt.hour


In [362]:
print(matches[['venue_code', 'opp_code', 'hour', 'day_code']].head())

   venue_code  opp_code  hour  day_code
0           0        10    12         5
1           1         3    16         6
2           0        16    16         6
3           1        19    15         5
4           1         2    15         5


In [369]:
# Create target column based on the 'result' column
matches['target'] = (matches['result'] == 'W').astype(int)

In [371]:
# Feature matrix and target
X = matches[predictors]
y = matches['target']

In [393]:
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# XGBoost training
import xgboost as xgb
model = xgb.XGBClassifier()
model.fit(X_train, y_train)

scale_pos_weight = len(y_train[y_train == 0]) / len(y_train[y_train == 1])

model = xgb.XGBClassifier(
    objective='binary:logistic',  # Binary classification
    eval_metric='logloss',       # Evaluation metric
    use_label_encoder=False,     # Suppress label encoding warning
    n_estimators=100,            # Number of boosting rounds
    learning_rate=0.1,           # Learning rate
    max_depth=3,                 # Maximum depth of a tree
    subsample=0.8,               # Subsample ratio of the training instances
    colsample_bytree=0.8         # Subsample ratio of columns when constructing each tree
)
param_grid = {
    'n_estimators': [100, 200],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2],
    'subsample': [0.7, 0.8, 0.9],
    'colsample_bytree': [0.7, 0.8, 0.9],
}

grid_search = GridSearchCV(
    estimator=model,
    param_grid=param_grid,
    scoring='accuracy',
    cv=3,
    verbose=1,
    n_jobs=-1
)

model.fit(X_train, y_train)
grid_search.fit(X_train, y_train)

# Evaluate accuracy
from sklearn.metrics import accuracy_score, classification_report
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2%}")
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

C:\Users\muham\AppData\Local\anaconda3\Lib\site-packages\xgboost\core.py:158: UserWarning: [00:31:02] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-0c55ff5f71b100e98-1\xgboost\xgboost-ci-windows\src\learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)


Fitting 3 folds for each of 162 candidates, totalling 486 fits
Accuracy: 61.70%

Classification Report:
              precision    recall  f1-score   support

           0       0.63      0.87      0.73       362
           1       0.57      0.25      0.35       249

    accuracy                           0.62       611
   macro avg       0.60      0.56      0.54       611
weighted avg       0.60      0.62      0.57       611



C:\Users\muham\AppData\Local\anaconda3\Lib\site-packages\xgboost\core.py:158: UserWarning: [00:31:06] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-0c55ff5f71b100e98-1\xgboost\xgboost-ci-windows\src\learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)
